app/src/main/java/com/example/util/simpletimetracker/wear/DomainAPI.kt

```diff
@@ -0,0 +1,76 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+package com.example.util.simpletimetracker.wear
+
+import com.example.util.simpletimetracker.domain.interactor.PrefsInteractor
+import com.example.util.simpletimetracker.domain.interactor.RecordTagInteractor
+import com.example.util.simpletimetracker.domain.interactor.RecordTypeInteractor
+import com.example.util.simpletimetracker.domain.interactor.RunningRecordInteractor
+import com.example.util.simpletimetracker.domain.mapper.AppColorMapper
+import com.example.util.simpletimetracker.domain.model.RunningRecord
+import com.example.util.simpletimetracker.wearrpc.Activity
+import com.example.util.simpletimetracker.wearrpc.CurrentActivity
+import com.example.util.simpletimetracker.wearrpc.Settings
+import com.example.util.simpletimetracker.wearrpc.SimpleTimeTrackerAPI
+import com.example.util.simpletimetracker.wearrpc.Tag
+
+class DomainAPI(
+    private val prefsInteractor: PrefsInteractor,
+    private val recordTypeInteractor: RecordTypeInteractor,
+    private val recordTagInteractor: RecordTagInteractor,
+    private val runningRecordInteractor: RunningRecordInteractor,
+    private val appColorMapper: AppColorMapper,
+) : SimpleTimeTrackerAPI {
+
+    override suspend fun queryActivities(): Array<Activity> {
+        return recordTypeInteractor.getAll().filter { recordType -> !recordType.hidden }
+            .map { recordType ->
+                val color = appColorMapper.mapToColorInt(recordType.color)
+                val hex = String.format("#%06X", (0xFFFFFF and color))
+                Activity(recordType.id, recordType.name, recordType.icon, hex)
+            }.toTypedArray()
+    }
+
+    override suspend fun queryCurrentActivities(): Array<CurrentActivity> {
+        return runningRecordInteractor.getAll().map { record ->
+            CurrentActivity(
+                record.id,
+                record.timeStarted,
+                record.tagIds.map { tagId ->
+                    val tag = recordTagInteractor.get(tagId)
+                    Tag(id = tag?.id ?: -1, name = tag?.name ?: "")
+                }.filter { it.id > 0 }.toTypedArray(),
+            )
+        }.toTypedArray()
+    }
+
+    override suspend fun setCurrentActivities(activities: Array<CurrentActivity>) {
```

Comment:
@kantahrek and I had this same discussion. We concluded that a single, declarative endpoint would be more reliable than separate start/stop endpoints.

The Watch and Phone constitute a distributed system, and the underlying `MessageClient` used for the communication layer explicitly **does not guarantee successful message delivery**. As a result, state-mutating requests need some mechanism for [idempotency](https://en.wikipedia.org/wiki/Idempotence) so they can be retried safely. The typical solution to this problem -- an [idempotency token](https://aws.amazon.com/builders-library/making-retries-safe-with-idempotent-APIs/#Reducing_client_complexity_with_idempotent_API_design) -- would require adding additional persistence on the phone.

By contrast, this declarative endpoint achieves idempotency without needing any persistence. When the watch doesn't receive a confirming response, it knows it can retry the request without risking creating unwanted activity records.

Reply:
I see your point, but still can't understand the difference. For example I click on activity on the watch, the watch sends `startActivity(id: Long)` event, there was a communication error, watch doesn't receive a confirming response and sends `startActivity(id: Long)` again, phone receives two events, if activity was started by the first one - second event will do nothing, if activity was not started by the first event - second will start it. Same goes for stopping activity. Will this not work?

Reply:
Can't pinpoint it but I have a gut feeling that it will cause more problems later on, guess we'll see)

Reply:
> Implementing a no-op for the presumed second startActivity requires the ability to distinguish between illegitimate duplicate requests and legitimate requests to restart an activity.

Can you explain a bit more, why it would require the ability to distinguish them?

Reply:
> > requests and legitimate requests to restart an activity.
> 
> Can you explain a bit more, why it would require the ability to distinguish them?

It depends on the semantics of `startActivity`.

- If `startActivity(id: Long)` always creates a new time-tracking record, then it's unsafe to automatically re-try. If the phone processes the first request, but the watch doesn't receive the response, automatic re-tries risk creating multiple time tracking entries.
	- This is how I viewed the method.
- If, in contrast, `startActivity`'s purpose is to ensure that the identified activity is running (starting it if necessary), then the method is idempotent.
	- I think this is how you've been viewing the method.

Reply:
> Can't pinpoint it but I have a gut feeling that it will cause more problems later on, guess we'll see)

The one gut concern I have with `setCurrentActivities` is that it clearly requires the watch to fully implement the semantics of some settings, particularly **multi-tasking**. A `start/stop` paradigm could handle the differing behaviors internally on the phone.

However, that potential optimization isn't universal for all settings. For example, the watch would have to re-implement logic to honor the settings for **Show record tag selection**.

Reply:
Thank you for the explanation! At the moment it is not possible to have two timers of the same activity, so you are right that my understanding was the second bullet point.

Also sorry for the confusion, by `startActivity(id: Long)` I meant `startActivity(all required information)`.

Reply:
For now, I think I'll leave it as `setCurrentActivities`. The other nice thing I found while implementing support for the phone's settings is that `setCurrentActivities` is atomic, while `start/stop` methods are not.